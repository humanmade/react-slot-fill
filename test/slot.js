/**
 * External dependencies
 */
import { isEmpty } from 'lodash';
import ReactTestRenderer from 'react-test-renderer';

/**
 * Internal dependencies
 */
import { Slot, Fill, Provider } from '../';

/**
 * WordPress dependencies
 */
import { Component } from '@wordpress/element';

class Filler extends Component {
	constructor() {
		super( ...arguments );

		this.state = {
			num: 1,
		};
	}
	render() {
		return [
			<button
				key="1"
				type="button"
				onClick={ () => this.setState( { num: this.state.num + 1 } ) }
			/>,
			<Fill name={ this.props.name } key="2">
				{ this.props.text || this.state.num.toString() }
			</Fill>,
		];
	}
}

describe( 'Slot', () => {
	it( 'should render empty Fills', () => {
		const tree = ReactTestRenderer.create(
			<Provider>
				<div>
					<Slot name="chicken" />
				</div>
				<Fill name="chicken" />
			</Provider>
		).toJSON();

		expect( tree ).toMatchSnapshot();
	} );

	it( 'should render a string Fill', () => {
		const tree = ReactTestRenderer.create(
			<Provider>
				<div>
					<Slot name="chicken" />
				</div>
				<Fill name="chicken">content</Fill>
			</Provider>
		).toJSON();

		expect( tree ).toMatchSnapshot();
	} );

	it( 'should render a Fill containing an element', () => {
		const tree = ReactTestRenderer.create(
			<Provider>
				<div>
					<Slot name="chicken" />
				</div>
				<Fill name="chicken">
					<span />
				</Fill>
			</Provider>
		).toJSON();

		expect( tree ).toMatchSnapshot();
	} );

	it( 'should render a Fill containing an array', () => {
		const tree = ReactTestRenderer.create(
			<Provider>
				<div>
					<Slot name="chicken" />
				</div>
				<Fill name="chicken">
					{ [ <span key="1" />, <div key="2" />, 'text' ] }
				</Fill>
			</Provider>
		).toJSON();

		expect( tree ).toMatchSnapshot();
	} );

	it( 'calls the functions passed as the Slotâ€™s fillProps in the Fill', () => {
		const onClose = jest.fn();

		const testInstance = ReactTestRenderer.create(
			<Provider>
				<Slot name="chicken" fillProps={ { onClose } } />
				<Fill name="chicken">
					{ ( props ) => {
						return (
							<button onClick={ props.onClose }>Click me</button>
						);
					} }
				</Fill>
			</Provider>
		).root;

		testInstance.findByType( 'button' ).props.onClick();

		expect( onClose ).toHaveBeenCalledTimes( 1 );
	} );

	it( 'should render empty Fills without HTML wrapper when render props used', () => {
		const tree = ReactTestRenderer.create(
			<Provider>
				<div>
					<Slot name="chicken">
						{ ( fills ) =>
							! isEmpty( fills ) && (
								<blockquote>{ fills }</blockquote>
							)
						}
					</Slot>
				</div>
				<Fill name="chicken" />
			</Provider>
		).toJSON();

		expect( tree ).toMatchSnapshot();
	} );

	it( 'should render a string Fill with HTML wrapper when render props used', () => {
		const tree = ReactTestRenderer.create(
			<Provider>
				<div>
					<Slot name="chicken">
						{ ( fills ) =>
							fills && <blockquote>{ fills }</blockquote>
						}
					</Slot>
				</div>
				<Fill name="chicken">content</Fill>
			</Provider>
		).toJSON();

		expect( tree ).toMatchSnapshot();
	} );

	it( 'should re-render Slot when not bubbling virtually', () => {
		const testRenderer = ReactTestRenderer.create(
			<Provider>
				<div>
					<Slot name="egg" />
				</div>
				<Filler name="egg" />
			</Provider>
		);

		expect( testRenderer.toJSON() ).toMatchSnapshot();

		testRenderer.root.findByType( 'button' ).props.onClick();

		expect( testRenderer.toJSON() ).toMatchSnapshot();
	} );

	it( 'should render in expected order', () => {
		const testRenderer = ReactTestRenderer.create(
			<Provider>
				<div key="slot">
					<Slot name="egg" />
				</div>
			</Provider>
		);

		testRenderer.update(
			<Provider>
				<div key="slot">
					<Slot name="egg" />
				</div>
				<Filler name="egg" key="first" text="first" />
				<Filler name="egg" key="second" text="second" />
			</Provider>
		);

		testRenderer.update(
			<Provider>
				<div key="slot">
					<Slot name="egg" />
				</div>
				<Filler name="egg" key="second" text="second" />
			</Provider>
		);

		testRenderer.update(
			<Provider>
				<div key="slot">
					<Slot name="egg" />
				</div>
				<Filler name="egg" key="first" text="first" />
				<Filler name="egg" key="second" text="second" />
			</Provider>
		);

		expect( testRenderer.toJSON() ).toMatchSnapshot();
	} );

	describe.each( [ false, true ] )(
		'bubblesVirtually %p',
		( bubblesVirtually ) => {
			it( 'should subsume another slot by the same name', () => {
				const testRenderer = ReactTestRenderer.create(
					<Provider>
						<div data-position="first">
							<Slot
								name="egg"
								bubblesVirtually={ bubblesVirtually }
							/>
						</div>
						<div data-position="second"></div>
						<Fill name="egg">Content</Fill>
					</Provider>
				);

				testRenderer.update(
					<Provider>
						<div data-position="first">
							<Slot
								name="egg"
								bubblesVirtually={ bubblesVirtually }
							/>
						</div>
						<div data-position="second">
							<Slot
								name="egg"
								bubblesVirtually={ bubblesVirtually }
							/>
						</div>
						<Fill name="egg">Content</Fill>
					</Provider>
				);

				expect( testRenderer.toJSON() ).toMatchSnapshot();

				testRenderer.update(
					<Provider>
						<div data-position="first"></div>
						<div data-position="second">
							<Slot
								name="egg"
								bubblesVirtually={ bubblesVirtually }
							/>
						</div>
						<Fill name="egg">Content</Fill>
					</Provider>
				);

				expect( testRenderer.toJSON() ).toMatchSnapshot();
			} );
		}
	);
} );
