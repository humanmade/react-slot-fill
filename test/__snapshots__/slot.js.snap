// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slot bubblesVirtually false should subsume another slot by the same name 1`] = `
Array [
  <div
    data-position="first"
  />,
  <div
    data-position="second"
  >
    Content
  </div>,
]
`;

exports[`Slot bubblesVirtually false should subsume another slot by the same name 2`] = `
Array [
  <div
    data-position="first"
  />,
  <div
    data-position="second"
  >
    Content
  </div>,
]
`;

exports[`Slot bubblesVirtually true should subsume another slot by the same name 1`] = `
Array [
  <div
    data-position="first"
  >
    <div />
  </div>,
  <div
    data-position="second"
  >
    <div />
  </div>,
]
`;

exports[`Slot bubblesVirtually true should subsume another slot by the same name 2`] = `
Array [
  <div
    data-position="first"
  />,
  <div
    data-position="second"
  >
    <div />
  </div>,
]
`;

exports[`Slot should re-render Slot when not bubbling virtually 1`] = `
Array [
  <div>
    1
  </div>,
  <button
    onClick={[Function]}
    type="button"
  />,
]
`;

exports[`Slot should re-render Slot when not bubbling virtually 2`] = `
Array [
  <div>
    2
  </div>,
  <button
    onClick={[Function]}
    type="button"
  />,
]
`;

exports[`Slot should render a Fill containing an array 1`] = `
<div>
  <span />
  <div />
  text
</div>
`;

exports[`Slot should render a Fill containing an element 1`] = `
<div>
  <span />
</div>
`;

exports[`Slot should render a string Fill 1`] = `
<div>
  content
</div>
`;

exports[`Slot should render a string Fill with HTML wrapper when render props used 1`] = `
<div>
  <blockquote>
    content
  </blockquote>
</div>
`;

exports[`Slot should render empty Fills 1`] = `<div />`;

exports[`Slot should render empty Fills without HTML wrapper when render props used 1`] = `<div />`;

exports[`Slot should render in expected order 1`] = `
Array [
  <div>
    first
    second
  </div>,
  <button
    onClick={[Function]}
    type="button"
  />,
  <button
    onClick={[Function]}
    type="button"
  />,
]
`;
